firewood/src/merkle.rs

```diff
@@ -1175,6 +1178,288 @@ impl<S: ShaleStore<Node> + Send + Sync> Merkle<S> {
     pub fn flush_dirty(&self) -> Option<()> {
         self.store.flush_dirty()
     }
+
+    pub fn get_iter<K: AsRef<[u8]>>(
+        &self,
+        key: Option<K>,
+        root: DiskAddress,
+    ) -> Result<MerkleKeyValueStream<'_, S>, MerkleError> {
+        // TODO: if DiskAddress::is_null() ...
+        Ok(MerkleKeyValueStream {
+            key_state: IteratorState::new(key),
+            merkle_root: root,
+            merkle: self,
+        })
+    }
+}
+
+enum IteratorState<'a> {
+    /// Start iterating at the beginning of the trie,
+    /// returning the lowest key/value pair first
+    StartAtBeginning,
+    /// Start iterating at the specified key
+    StartAtKey(Vec<u8>),
+    /// Continue iterating after the given last_node and parents
+    Iterating {
+        last_node: ObjRef<'a>,
+        parents: Vec<(ObjRef<'a>, u8)>,
+    },
+}
+impl IteratorState<'_> {
+    fn new<K: AsRef<[u8]>>(starting: Option<K>) -> Self {
+        match starting {
+            None => Self::StartAtBeginning,
+            Some(key) => Self::StartAtKey(key.as_ref().to_vec()),
+        }
+    }
+}
+
+// The default state is to start at the beginning
+impl<'a> Default for IteratorState<'a> {
+    fn default() -> Self {
+        Self::StartAtBeginning
+    }
+}
+
+/// A MerkleKeyValueStream iterates over keys/values for a merkle trie.
+/// This iterator is not fused. If you read past the None value, you start
+/// over at the beginning. If you need a fused iterator, consider using
+/// std::iter::fuse
+pub struct MerkleKeyValueStream<'a, S> {
+    key_state: IteratorState<'a>,
+    merkle_root: DiskAddress,
+    merkle: &'a Merkle<S>,
+}
+
+impl<'a, S: shale::ShaleStore<node::Node> + Send + Sync> Stream for MerkleKeyValueStream<'a, S> {
+    type Item = Result<(Vec<u8>, Vec<u8>), api::Error>;
+
+    fn poll_next(
+        mut self: std::pin::Pin<&mut Self>,
+        _cx: &mut std::task::Context<'_>,
+    ) -> Poll<Option<Self::Item>> {
+        // Note that this sets the key_state to StartAtBeginning temporarily
+        let found_key = match std::mem::take(&mut self.key_state) {
+            IteratorState::StartAtBeginning => {
+                let root_node = self
+                    .merkle
+                    .get_node(self.merkle_root)
+                    .map_err(|e| api::Error::InternalError(e.into()))?;
+                let mut last_node = root_node;
+                let mut parents = vec![];
+                let leaf = loop {
+                    match last_node.inner() {
+                        NodeType::Branch(branch) => {
+                            if let Some((leftmost_position, leftmost_address)) = branch
+                                .children
+                                .iter()
+                                .enumerate()
+                                .find(|&addr| addr.1.is_some())
```

Comment:
I don't like this because it will get undone once we change DiskAddress in the node [as in this PR](https://github.com/ava-labs/firewood/pull/341). I want to just remove the unwrap later (or land #341 first).

Reply:
Fixed per discussion later on

Reply:
Applied this